### ISSUE TITLE
[Transport endpoint is not connected], mounts and temporary mounts left in inconsistent state.

### ISSUE THREAD

**Issue by Francesco-n-dev** [mxmlnkn/ratarmount#131]
Hi there!
Let me start this conversation by congratulating you on this project and thanking you for sharing it with the whole community!

Unfortunately we are experiencing some issues while using `ratarmount` to mount a .tar.gz archive as a filesystem.

Some context information first:

Os:
```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye
```
Hardware:
```bash
$ cat /proc/meminfo
...
MemTotal:       16392436 kB
...
```
```bash
$ cat /proc/cpuinfo
...
cpu MHz		: 2249.998
cache size	: 512 KB
cpu cores	: 2
...
```
```bash
$ df -T
Filesystem          Type      1K-blocks     Used     Available   Use%  Mounted on
...
/dev/sda1           ext4       20401796  5692496  13755808    30%     /
storage0             zfs       101599360  1022688 100576672   2%       /myfolder
...
```
Notice that mounts and temporary mounts goes in different filesystems, mounted folders will be in `/myfolder` which is a `zfs` type filesystem and `/` where the `/tmp/.ratarm..` mounts goes is a `ext4` filesystem.

Ratarmount version:
```bash
ratarmount --version

ratarmount 0.14.1
ratarmountcore 0.6.3

System Software:

Python 3.12.2
FUSE 2.9.2
libsqlite3 3.45.1

Compression Backends:

indexed_bzip2 1.6.0
rapidgzip 0.10.5
indexed_gzip 1.8.7
xz 0.4.0
indexed_zstd 1.1.3
rarfile 4.1
```
We installed ratarmount by downloading the app image from https://github.com/mxmlnkn/ratarmount/releases/download/v0.14.0/ratarmount-0.14.0-x86_64.AppImage

Commands we use to mount the .tar.gz archive:
```bash
ratarmount -w /myfolder/mount_writes --fuse "uid=33,gid=33" /myfolder/mytargzfile.tar.gz /myfolder/mount_folder
```
Command we use to unmount the archive:
```bash
ratarmount -u /myfolder/mount_folder
```
**Context**:
- In our environment we have a jenkins service which we use to schedule and execute a large amount of jobs, these jobs are executed on a group of nodes that can vary over time.
- Every job has its own folder with its own .tar.gz so that jobs don't have to share any kind of resource except memory and cpu.
- Every job has to mount the .tar.gz before executing some routines, mostly php scripts, in the mounted folder and then unmount it before the job ends.
- On every jenkins swarm node a maximum of 19 .tar.gz mounts can happen and exists at the same time on different folders and using the .tar.gz with the same contents but copied and located in different folders too.
- Some Jobs needs to write files inside the mounted folder.

It is not uncommon for jobs to end abruptly, in that case we found out jenkins will terminate the child processes of the job which includes of course ratarmount.

**Problems**:

1) On some jenkins swarm nodes we noticed the increasing amount of temporary mounts left in an inconsistent state.

Executing the command `df -h`, other then the healthy filesystems, we get errors like these:
```bash
$ df -h
...
/tmp/.mount_ratarmI3YrXb: Transport endpoint is not connected
/tmp/.mount_ratarmoBkgRN: Transport endpoint is not connected
...
```
and by listing all the contents of /tmp with ls command we observe this:
```bash
$ ls -lash /tmp
...
? d?????????   ? ?                 ?                    ?            ? .mount_ratarmI3YrXb
? d?????????   ? ?                 ?                    ?            ? .mount_ratarmoBkgRN
...
```
As we start to investigate we didn't find any hanging process that could be responsible for the corrupted mounts.
As next step we tried to unmount them with `ratarmount` unsuccessfully:
```bash
$ ratarmount -u /tmp/.mount_ratarmI3YrXb
[Error] The given path to unmount (/tmp/.mount_ratarmI3YrXb) must be a mount point!
```
then we tried with `umount` also without success:
```bash
$ umount -l  /tmp/.mount_ratarmI3YrXb
umount: /tmp/.mount_ratarmI3YrXb: Transport endpoint is not connected
```
with `fusermount` we finally made it:
```bash
$ fusermount -u /tmp/.mount_ratarmI3YrXb
```
**Reproduce**:
Unfortunately we weren't able to reproduce this behaviour consistently but it's seems that trying to start multiple mounts at the same time, even if the .tar.gz files and folders are all different, could somehow lead to this issue after the job ends.
Manually killing the jobs, processes, including ratarmount ones, or jenkins swarm service never lead to the inconsistent state.

---

 2) On some jenkins swarm nodes we noticed mounted folders left in an inconsistent state.

Sometimes we noticed we are left with some mounted folder in an inconsistent state similiar to the problem 1.
Because every job will run multiple times inside their own folder the first step is to try to create, if it doesn't already exists, the folder on which the .tar.gz will be mounted, if the folder already exists it means some previous job instances already created it.
The command below shows what happens when the job start trying to access the inconsistent folder:
```bash
$ mkdir -p /myfolder/mount_folder
mkdir: cannot stat ‘mount_folder’: Transport endpoint is not connected
```
Like in the problem 1 we observed the same inconsistencies but this time those involves the mounted folder instead.
Investigating the inconsistent `/myfolder/mount_folder`:
```bash
$ df -h
...
/myfolder/mount_folder: Transport endpoint is not connected
...
```
```bash
$ ls -lash /myfolder
...
27M -rw-r--r--  1 33 33 27M Mar 27 10:42  mytargzfile.tar.gz
? d?????????  ? ?        ?          ?            ? mount_folder
512 drwxr-xr-x  3 33 33   5 Mar 27 10:44 mount_writes
...
```
The only way to unmount the filesystem is, as in the problem 1, using `fusermount`.

**Reproduce**:
We were able to reproduce this behaviour quite consistently by spawning jobs that consumes memory that lead the os to invoke oom-kill on the jenkins swarm service leaving the mounted folders of some jobs in an inconsistent state.

---

We acknowledge that things like `oom-kill` or any sort of process interruption could lead to inconsistent statuses, we would like to know if someone else experienced this and eventually which solution or strategy might help us.

We also don't have any clue of why the only way to unmount the inconsistent filesystems is by using `fusermount`.
Maybe the fact that we have multiple processes trying to mount the filesystems on different paths but at the same time could lead to some kind of problem? I won't exclude we are missing some option or best practices about ratarmount or mounting archives as filesystems in general.

Big thanks in advance for everyone that will partecipate to this issue!

**Comment by mxmlnkn** [comment]
This is a very common issue with FUSE in general: [[0]](https://stackoverflow.com/questions/24966676/transport-endpoint-is-not-connected) [[1]](https://stackoverflow.com/questions/16002539/fuse-error-transport-endpoint-is-not-connected).

What happens is that the FUSE mount point is not properly unmounted while the actual process that provides the contents for the mount point, the "transport endpoint" I assume, in this case: ratarmount, does not exist anymore, i.e., the FUSE kernel module(?) is not able to query the user-space process to ask it what files should be shown for the mount point.

This issue can be reproduced easily when uncleanly killing ratarmount with `-9` (SIGKILL) instead of the default SIGTERM:

```bash
mkdir folder
ratarmount folder mounted
ls -a mounted  # . ..
kill -9 $( pgrep ratarmount | tail -1 )
ls -a mounted  # ls: cannot access 'mounted': Transport endpoint is not connected
ls -l  # d????????? ? ?           ?              ?            ? mounted
```

If there is something that is killing ratarmount (oom-killer for example), then it should be configured to first try to close ratarmount cleanly without -9 and only if that does not succeed after a timeout should it use -9.

It might also happen when ratarmount or one of the backends crashes hard, e.g., a segfault. If that is the case, then I'd need some reproducing minimal example to try and find and fix the underlying issue.

About umount vs fusermount, those are intended for different things, so I'm not surprised that umount does not work. For FUSE mounts, only fusermount should be used, as far as I know.

**Comment by Francesco-n-dev** [comment]
Thanks for the quick reply!

We tried to reproduce the issue as you did and infact we noticed that killing the `ratarmount` process with `-9` produce the `Transport endpoint is not connected` on the temporary mounts, (/tmp/.ratarm...), while the process `/tmp/.mount_ratarm.../opt/python3.12..` is still alive, then if we try to kill this process too with `-9` it will produce the `Transport endpoint is not connected` on the mounted folder too.

Thats awesome!

Do you always suggest using fusermount to unmount the folders? or is there no difference? of course, unless there is an inconsistent situation, in which case fusermount is the only working option.

We tried to find if there is any way we can see the outputs of the ratarmount processes to get some insights of the kills happening, but we didn't found any, there is some way we can achieve this? In this way we will be able to understand without a doubt who and why makes the killing, this could help us to make things as gracefully as possible for the ratarmount processes.

Big thanks again!

**Comment by mxmlnkn** [comment]
> Do you always suggest using fusermount to unmount the folders? or is there no difference? of course, unless there is an inconsistent situation, in which case fusermount is the only working option.

My suggestion would be to use `ratarmount -u`, but I see now that it didn't work for you either.
 - [ ] This is a bug and should be fixed.

> We tried to find if there is any way we can see the outputs of the ratarmount processes to get some insights of the kills happening, but we didn't found any, there is some way we can achieve this?

I don't know. How are you starting ratarmount? When simply calling ratarmount, it moves itself into the background. This hides any error output. I'd advise to start it like this: `nohup ratarmount -d 3 -f ... &> ratarmount-$( date +%s ).log &`. `-f` prevents ratarmount from moving itself into the background, `-d 3` increases the logging to the maximum, ad `nohup ... &> ...log &` logs everything into a file and moves ratarmount to the background. It is possibly that this might already show some interesting errors. I would also track the memory usage because it can grow quite large for the gzip backend. Alternatively, parallelization could be limited with `-P 1` to also limit memory usage.

**Comment by Francesco-n-dev** [comment]
> I don't know. How are you starting ratarmount?

Our pipeline lives in a /bin/bash shell instance in which at the beginning we mount the folder like so:

```bash
# start pipeline
$ cmd0
$ cmd1
...
$ ratarmount -w /myfolder/mount_writes --fuse "uid=33,gid=33" /myfolder/mytargzfile.tar.gz /myfolder/mount_folder
...
$ cmdn
...
$ ratarmount -u /myfolder/mount_folder
# end pipeline
```

We tried to catch some output as you suggested and as the process "python" were being killed with `-9` signal this row has been produced:
```python
/tmp/.mount_ratarm4izltL/usr/bin/python3.12: line 25: 729233 Killed                  "$APPDIR/opt/python3.12/bin/python3.12" "$@"
```
killing the "ratarmount" process doesn't produce any output whatsoever.

At this point and by all the analysis we made we are pretty confident that the incident is not caused by ratarmount rather by other processes that causes oom-kill to intervene and leave mounts in an inconsistent state.

We will also try to use `-P` to see what are the effects on performance and resource consumption, nice catch!

> My suggestion would be to use ratarmount -u, but I see now that it didn't work for you either.

Infact if one process of the two that the command ratarmount spawn is being killed it wont bring the other down, killing with `-9` the process `ratarmount -w write_files mytargz.tar.gz mount_folder ...` wont bring the `/tmp/.mount_ratarm.../opt/python3.12..` process down creating the `/tmp/.mount_ratarm...` inconsistency, viceversa the kill signal `-9` on the "python process" bring the other one down and as we have seen this leaves the mounted folder inconsistent.

Maybe could help to somehow link them in both ways so that if the `ratarmount -w ` is being killed the other goes down too, but i'm just speculating, In this way there will be no instances where fusermount is the only option to resolve one of the two possible inconsistencies.

On my side feel free to close this issue, if there is not already and you feel that is needed we might think to create another issue about the `ratarmount -u` not working properly on inconsistent situations; Imho in some ways it makes sense that it doesn't work properly because the inconsistency is no longer in the hands of ratarmount itself but in the hand of the other layers.

Thanks again!

**Comment by mxmlnkn** [comment]
> Infact if one process of the two that the command ratarmount spawn is being killed it wont bring the other down

 - [ ] I didn't understand fully why you were mentioning a mount point in /tmp and two processes, but it makes sense. You did say that you were using the AppImage. The AppImage executable first mounts itself to temp and then calls the actual ratarmount process. I wasn't fully aware of the implications. Maybe there is something I can do to the AppImage shell script to detect if the started process crashes.

As for `ratarmount -u`, I simply [removed](https://github.com/mxmlnkn/ratarmount/commit/4f765ce3366e750691c9c0ea550a95baf671149f) the `os.path.ismount` check. This fixes the issue for me. It seems that [CPython's ismount](https://github.com/python/cpython/issues/96328) has some rough edges.

**Comment by mxmlnkn** [comment]
The following deliberations are mostly for my own understanding and for posterity.

    21 ├─ ratarmount-0.14.1-x86_64.AppImage│./ratarmount-0.14.1-x86_64.AppImage silesia.tar.gz mounted
    22 │  └─ ratarmount-0.14.1-x86_64.AppImage│./ratarmount-0.14.1-x86_64.AppImage silesia.tar.gz mounted
    23 ├─ /tmp/.mount_ratarmSv5sVA/opt/python3.12/bin/python3.12 -u -I -m ratarmount silesia.tar.gz mounted

Interpretation:

 - Process 21 is the initially started AppImage.
 - Either process 21 or 22 provides the `/tmp/.mount_ratarmSv5sVA` FUSE mount. I'm not sure.
 - Process 23 is the actual ratarmount process that provides the `mounted` FUSE mount.
   - It is not in the process hierarchy of 21 or 22 because it daemonizes itself, I think.
     This behavior should be preventable with `-f`.

Behavior:

 - Killing 22 also makes 21 exit while the AppImage mount point at `/tmp/.mount` persists in broken state.
   It is unexpected that process 23 and even the mount point `mounted` work completely fine.
   It seems like all necessary files such as python3.12 are mapped into memory so that the broken `/tmp.mount` point does not matter.
 - Killing process 23 also kills 21 and 22 while the provided FUSE mount in `mounted` persists.
   Process 23 cannot do anything to correctly unmount `mounted`.
   Process 21 or 22, either of which should run `entrypoint.sh` might be able to do something but it would require
   full duplicated argument parsing and other uncertain magic.
   Maybe Process 23 could launch an independent watcher-process that watches Process 23 and calls fusermount.
   It must not be a child of Process 22 or else it would get killed without any trap being called too, I assume.
   The downside would be yet another background process...

Starting ratarmount without letting it fork itself into the background like so:

    nohup ./ratarmount-0.14.1-x86_64.AppImage -f silesia.tar.gz mounted &> ratarmount-$( date +%s ).log &

results in this process hierarchy looks like this:

    21 │ ├─ ratarmount-0.14.1-x86_64.AppImage│./ratarmount-0.14.1-x86_64.AppImage silesia.tar.gz mounted
    22 │    └─ ratarmount-0.14.1-x86_64.AppImage│./ratarmount-0.14.1-x86_64.AppImage silesia.tar.gz mounted
    1  ├─ bash
    23 │  └─ /tmp/.mount_ratarmSv5sVA/opt/python3.12/bin/python3.12 -u -I -m ratarmount silesia.tar.gz mounted

In this case, the AppImage processes have forked into the background and process 23 is a child process of the calling bash process.
Unfortunately, this doesn't change much. The AppImage and ratarmount processes are still independent.
I'm actually interested how Process 21 or 22 notice that 23 quits in order to automatically unmount `/tmp/.mount`.
I might have to look into the AppImage source code for that.

In conclusion, I think that for now some kind of cleanup script that looks for broken mount points and tries to unmount them, would be the easiest solution, and in the first place it should be avoided to kill or crash the processes this hard in the first place.

I'll close this issue for now as you suggested. If you have any other issues, feel free to open another one or reply here.

**Comment by mxmlnkn** [comment]
Btw, I'm currently preparing the new AppImages for ratarmount 0.15 with libarchive support. And the AppImage test runtimes become unbearable with the plethora of newly added tests. To save the AppImage decompression overhead and all the trouble it brings you with the extra background processes and the additional FUSE mount, you can also simply extract the AppImage and call the AppRun script directly! It is one installation step more, but if you use it frequently, it should be worth it.

```
time ./ratarmount-0.14.2-x86_64.AppImage  # 1.2s
./ratarmount-0.14.2-x86_64.AppImage --appimage-extract
mv squashfs-root "$HOME/bin/ratarmount-0.14.2"
alias ratarmount="$HOME/bin/ratarmount-0.14.2/AppRun"
time ratarmount  # 0.4s
```

**Comment by Francesco-n-dev** [comment]
Hello!
The upgrade to 0.15.0 is very much appreciated, good job!!!
We promptly tested the 0.15.0 implementation following your steps and it works like a charm! The fact that we have to manage less processes makes our environment more clean and manageable!

With the 0.14.x using the AppImage without extraction, as you know, we had to create a little garbage collector that parses the `/tmp/` folders and check if the mounts, (with `.mount_ratar` in their names), are healthy, if not unmount them with fusermount.

We also noticed that alot of `/tmp/.mount_ratarm..` folders were left behind, healthy but empty, our garbage collector handles that too by removing them because it seems they are no longer needed.

Anyways, with the steps you provided garbage collection seems unnecessary, no more leftovers in `/tmp/`! Thank you very much!

Giving our 2 cents here, for those who still have to clean mounts programmatically, based on the contents in `/tmp/`:
```bash
#!/bin/bash
shopt -s nullglob dotglob
ratarm_tmp_mounts=(/tmp/.mount_ratarm*)
echo "total ratarmount tmp directories: ${#ratarm_tmp_mounts[@]}"
tot_empty=0
tot_broken=0
tot_busy=0
if [ "${#ratarm_tmp_mounts[@]}" -gt 0 ]; then
    for ratarm_tmp_mount in "${ratarm_tmp_mounts[@]}"; do
        # try to count elements in ratarmount tmp directory, if error occurs, (broken mountpoint), returns 0
        if [ "$(find "$ratarm_tmp_mount" -mindepth 1 2>/dev/null | wc -l)" -eq 0 ]; then
            # try to remove directory, if error is a broken mountpoint
            if rm -rf "$ratarm_tmp_mount" 2>/dev/null; then
                tot_empty=$((tot_empty + 1))
                echo "removed empty folder [$ratarm_tmp_mount]"
            else
                tot_broken=$((tot_broken + 1))
                fusermount -u "$ratarm_tmp_mount"
                rm -rf "$ratarm_tmp_mount"
                echo "broken mountpoint, unmounted and removed: [$ratarm_tmp_mount]"
            fi
        else
            tot_busy=$((tot_busy + 1))
            echo "ratarmount tmp directory not empty: [$ratarm_tmp_mount]"
        fi
    done
    echo "done, empty directories: [$tot_empty] broken directories: [$tot_broken] busy directories: [$tot_busy] on total: [${#ratarm_tmp_mounts[@]}]"
fi
shopt -u nullglob dotglob
```
Below a garbage collector based on mounts that has `ratarmount` in their directory name, this is just an extra but it might be useful for someone =)
```bash
#!/bin/bash
shopt -s nullglob dotglob
ratarm_mounts=($(mount | sed '/^#/d' | sed '/^$/d' | sed 's/  */ /g' | cut -d" " -f3 | grep "mounttest"))
echo "total ratarmount mounts: ${#ratarm_mounts[@]}"
tot_broken=0
tot_ok=0
if [ "${#ratarm_mounts[@]}" -gt 0 ]; then
    for ratarm_mount in "${ratarm_mounts[@]}"; do
        # test if mountpoint is healthy
        if [ "$(mountpoint "$ratarm_mount" 2>/dev/null)" ]; then
            tot_ok=$((tot_ok + 1))
            echo "ratarmount mount healthy: [$ratarm_mount]"
        else
            tot_broken=$((tot_broken + 1))
            fusermount -u "$ratarm_mount"
            echo "broken mountpoint, unmounted: [$ratarm_mount]"
        fi
    done
    echo "done, broken mounts: [$tot_broken] busy mounts: [$tot_ok] on total: [${#ratarm_mounts[@]}]"
fi
shopt -u nullglob dotglob
```
Be weary! If the mount is in an inconsistent state and a process still has some work to do on the mounted directory the command `fusermount -u` returns error because it still see the mount as busy, this should be an edge case though since the process should immediately fail as the files he uses cease to exists.

Big Thanks again!
